client/lib/user/user.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * External Dependencies
+ */
+import { EventEmitter } from 'events';
+import { Language } from '../../languages';
+import { GravatarOptions } from './gravatar';
+import { URL } from '../../types';
+
+/**
+ * Takes an optional callback function. If no callback is passed in, it will return a promise.
+ **/
+export interface RequestHandler {
+	( fn: ( err: unknown, res: unknown, headers: unknown ) => unknown ): unknown;
+	(): Promise< unknown >;
+}
+
+export interface User extends EventEmitter {
+	initialize: () => Promise< void >;
+	clearStoreIfChanged: () => void;
+	get: () => UserData;
+	fetch: () => Promise< void >;
+	handleFetchFailure: ( error: Error ) => void;
+	handleFetchSuccess: ( userdata: UserData ) => void;
+	getLanguage: () => Language | void;
```

Comment:
cc: @sirreal do you have any thoughts on returning `void` vs. returning `undefined` here? to me it seems like `Language | undefined` would make more sense because the function is specifically returning _the value_ `undefined` vs. the function is not intended to return a value.

@roo2 you might consider the nuance here and see if you have any thoughts. I think this is the first time I've come across this situation

Reply:
hmm I can't really decide what I think is better here, I couldn't find examples of optional return types yet in our code and couldn't really find a clear answer online.. I think as long as we're consistent either is ok, void is usually used for return types but undefined is usually used for optional types. 
I like `void` better because it's a simple rule, use void for return types and undefined everywhere else.

Reply:
in this case it's not even that we're doing `return;` but rather we are explicitly returning _the value_ `undefined` which is JS nonsense

I don't think the rule should be laid out on which is used in a return value but rather the purpose for which each one is used. `void` feels very related to functions returning no value. `undefined` connotes the _presence_ of the value `undefined`, and `?` or `Partial` connotes the _absence_ of a value of the given name.

let's see if Jon chimes in

Reply:
This is a bit funny, but as far as I can tell [`void`](https://www.typescriptlang.org/docs/handbook/basic-types.html#void) is effectively

```ts
type void = undefined
```

if `strictNullChecks` is enabled, or

```ts
type void = null | undefined;
```

otherwise.

---

Thinking of `void` as the absence of value seems appropriate. Therefore, I tend to avoid it in type unions:

```ts
// Nothing meaningful will come out of this callback
type Callback = () => void;
// This returns values, one of which is `undefined`:
type Head<T> = (arr: T[]) => T | undefined
```

My rule of thumb is that `void` doesn't appear in unions with other types 🙂 so here I'd prefer `Language | undefined` 👍